pkg/validating-webhooks/agentinstall/v1beta1/infraenv_admission_hook.go

```diff
@@ -0,0 +1,194 @@
+	}
+}
+
+// shouldValidate explicitly checks if the request should be validated. For example, this webhook may have accidentally been registered to check
+// the validity of some other type of object with a different GVR.
+func (a *InfraEnvValidatingAdmissionHook) shouldValidate(admissionSpec *admissionv1.AdmissionRequest) bool {
+	contextLogger := log.WithFields(log.Fields{
+		"operation": admissionSpec.Operation,
+		"group":     admissionSpec.Resource.Group,
+		"version":   admissionSpec.Resource.Version,
+		"resource":  admissionSpec.Resource.Resource,
+		"method":    "shouldValidate",
+	})
+
+	if admissionSpec.Resource.Group != infraEnvGroup {
+		contextLogger.Debug("Returning False, not our group")
+		return false
+	}
+
+	if admissionSpec.Resource.Version != infraEnvVersion {
+		contextLogger.Debug("Returning False, it's our group, but not the right version")
+		return false
+	}
+
+	if admissionSpec.Resource.Resource != infraEnvResource {
+		contextLogger.Debug("Returning False, it's our group and version, but not the right resource")
+		return false
+	}
+
+	// If we get here, then we're supposed to validate the object.
+	contextLogger.Debug("Returning True, passed all prerequisites.")
+	return true
+}
+
+// validateUpdate specifically validates update operations for AgentClusterInstall objects.
+func (a *InfraEnvValidatingAdmissionHook) validateUpdate(admissionSpec *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse {
+	contextLogger := log.WithFields(log.Fields{
+		"operation": admissionSpec.Operation,
+		"group":     admissionSpec.Resource.Group,
+		"version":   admissionSpec.Resource.Version,
+		"resource":  admissionSpec.Resource.Resource,
+		"method":    "validateUpdate",
+	})
+
+	newObject := &v1beta1.InfraEnv{}
+	if err := a.decoder.DecodeRaw(admissionSpec.Object, newObject); err != nil {
+		contextLogger.Errorf("Failed unmarshaling Object: %v", err.Error())
+		return &admissionv1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+				Message: err.Error(),
+			},
+		}
+	}
+
+	// Add the new data to the contextLogger
+	contextLogger.Data["object.Name"] = newObject.Name
+
+	oldObject := &v1beta1.InfraEnv{}
+	if err := a.decoder.DecodeRaw(admissionSpec.OldObject, oldObject); err != nil {
+		contextLogger.Errorf("Failed unmarshaling OldObject: %v", err.Error())
+		return &admissionv1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+				Message: err.Error(),
+			},
+		}
+	}
+
+	if oldObject.Spec.ClusterRef != nil && (newObject.Spec.ClusterRef.Name != oldObject.Spec.ClusterRef.Name || newObject.Spec.ClusterRef.Namespace != oldObject.Spec.ClusterRef.Namespace) {
+		message := fmt.Sprintf("Attempted to change newObject.Spec.ClusterMetadata.ClusterID which is immutable after install started)")
```

Comment:
Please update the message:
Attempted to change Spec.ClusterRef which is immutable after InfraEnv creation.